### PR TITLE
[JENKINS-42728] Updating view with CLI using input of a different view type should fail

### PR DESCRIPTION
**Pull request by scoheb**
[FIXES JENKINS-42728]

# Description

See [JENKINS-42728](https://issues.jenkins-ci.org/browse/JENKINS-42728).

# Changelog Entry

* Fail with error when updating view with CLI using input of a different view type. In such case the view should be deleted and then created using other CLI commands.

### CODE REVIEW

**Comment by scoheb** [comment]
@jenkinsci/code-reviewers

File: core/src/main/java/hudson/model/View.java [hunk 1]
```diff
@@ -1186,12 +1186,16 @@ public void updateByXml(Source source) throws IOException {
...
 
         // try to reflect the changes by reloading
-
         try (InputStream in = new BufferedInputStream(new ByteArrayInputStream(out.toString().getBytes("UTF-8")))){
             // Do not allow overwriting view name as it might collide with another
             // view in same ViewGroup and might not satisfy Jenkins.checkGoodName.
             String oldname = name;
-            Jenkins.XSTREAM.unmarshal(new Xpp3Driver().createReader(in), this);
+            Object o = Jenkins.XSTREAM.unmarshal(new Xpp3Driver().createReader(in), this);
+            if (o!=this) {
```
Comment (olivergondza) [inline core/src/main/java/hudson/model/View.java]: Not `!o.getClass().equals(getClass())`?

Comment (scoheb) [inline core/src/main/java/hudson/model/View.java]: I was basing myself on the "tried, tested and true":

https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AbstractItem.java#L673

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/View.java]: +1 to @olivergondza 's proposal. Though I am not sure if it is going to be safe for the data migration logic like [class renaming](https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility). Maybe it worth doing some direct tests for this use-case.

`o != this` seems to be pretty nangerous

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I would rather prefer to have explicit class type checks, because I do not understand how the  current implementation may work reliably

**Comment by scoheb** [comment]
@oleg-nenashev  @olivergondza 

I was not sure where to put 

> TestView.java

Let me know if you think we need more tests.

File: test/src/test/java/hudson/cli/UpdateViewCommandTest.java [hunk 2]
```diff
@@ -66,6 +68,38 @@
...
+    @Test public void updateViewWithRenamedClass() throws Exception {
+        ListView tv  = new ListView("tView");
+        j.jenkins.addView(tv);
+        j.jenkins.XSTREAM2.addCompatibilityAlias("org.acme.old.Foo", ListView.class);
+        final CLICommandInvoker.Result result = command
+                .authorizedTo(View.READ, View.CONFIGURE, Jenkins.READ)
+                .withStdin(this.getClass().getResourceAsStream("/hudson/cli/testview-foo.xml"))
+                .invokeWithArgs("tView");
+
+        assertThat(result, failedWith(0));
```
Comment (olivergondza) [inline test/src/test/java/hudson/cli/UpdateViewCommandTest.java]: `succeeded()` or `succeededSilently()`.

**Review by olivergondza — CHANGES_REQUESTED** [review]

**Comment by oleg-nenashev** [comment]
Will try to take another look ASAP

File: test/src/test/java/hudson/cli/UpdateViewCommandTest.java [hunk 3]
```diff
@@ -66,6 +68,38 @@
         assertThat(result.stderr(), containsString("ERROR: user is missing the View/Configure permission"));
     }
 
+    /**
+     * This test shows that updating a view using an XML that will be
+     * converted by XStream via an alias will rightfully succeed.
+     *
+     * @throws Exception
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/cli/UpdateViewCommandTest.java]: Better to have a description defined. Otherwise Javadoc engine may grumble about that

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me

**Comment by oleg-nenashev** [comment]
@olivergondza WDYT?

File: test/src/test/java/hudson/cli/UpdateViewCommandTest.java (hunk 3 above)
Comment (olivergondza) [inline test/src/test/java/hudson/cli/UpdateViewCommandTest.java]: Or avoid javadoc (`/*`) for tests, it is not generated anyway. `@throws Exception` on tests is quite pointless.

**Review by olivergondza — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Ready to merge. Just to be safe, let's re-run CI after the security release on Wednesday before merging

**Pull request merged**